src/main/java/com/rackspacecloud/metrics/ingestionservice/influxdb/GCLineProtocolBackupService.java

```diff
@@ -0,0 +1,115 @@
+    private final String cloudOutputBucket;
+
+    // Internal methods of this class attempt to talk to the Cache proxy
+    // instead of the GCLineProtocolBackupService object
+    @Resource
+    private LineProtocolBackupService self;
+
+    // To autowire Storage in prod we'll need to get an account and point
+    // GOOGLE_APPLICATION_CREDENTIALS to the json creds file
+    // for a service account
+    // as per https://www.baeldung.com/java-google-cloud-storage
+    @Autowired
+    public GCLineProtocolBackupService(Storage storage, @Value("${backup.gcs-backup-bucket}") String cloudOutputBucket) {
+        Assert.notNull(storage, "Storage must not be null");
+        Assert.notNull(cloudOutputBucket, "The output bucket must not be null");
+        this.storage = storage;
+        this.cloudOutputBucket = cloudOutputBucket;
+    }
+
+    /**
+     * We want data to be stored in the following format: /[profile-specific
+     * bucket]/[instance name]/[database name]/[retention policy
+     * name]/[YYYYMMDD]/[UUID of file].gz
+     *
+     * File is line-protocol, gzipped, size may require some testing
+     *
+     * @param fileName the full name of the file we are writing to
+     * @return the cached reference to the gzip output stream for that blob
+     */
+    @Cacheable(cacheNames = "lineProtocolBackupWriter")
+    public GZIPOutputStream getBackupStream(String fileName) throws IOException {
+        Assert.notNull(fileName, "fileName must not be null");
+        Assert.notNull(cloudOutputBucket, "bucket name was not injected properly");
+        Assert.notNull(storage, "storage was not injected properly");
+        BlobId blobId = BlobId.of(cloudOutputBucket, fileName);
+        BlobInfo blobInfo = BlobInfo.newBuilder(blobId).setContentType("application/gzip").build();
+        // enable flushing the compressor
+        return new GZIPOutputStream(Channels.newOutputStream(storage.writer(blobInfo)), true);
+    }
+
+    @Override
+    public void writeToBackup(String payload, String instance, String database, String retentionPolicy)
+            throws IOException {
+        self.getBackupStream(getBackupFilename(payload, instance, database, retentionPolicy)).write(payload.getBytes());
+    }
+
+    public static final RemovalListener removalListener = (RemovalListener<String, GZIPOutputStream>) (key, value, cause) -> {
+        try {
+            value.finish();
+            value.close();
+        } catch (IOException e) {
+            // this will be eliminated and logged automatically
+            throw new RuntimeException(e);
+        }
+    };
+
+    private static final Pattern payloadTimestampPattern = Pattern.compile(" ([0-9]*)$");
+
+    private static long parseTimestampFromPayload(String payload) {
+        Matcher m = payloadTimestampPattern.matcher(payload);
+        m.find();
+        return Integer.valueOf(m.group(1));
+    }
+
+    private static final LocalUUID uuidGenerator = new LocalUUID();
```

Comment:
The `Whitebox.setInternalState` in the unit test

https://github.com/racker/ceres-ingestion-service/pull/9/files#diff-9b2b1d95218b932dc5508feda3b2277fR66

seems to raise a flag that this `uuidGenerator` should simply be autowired into the service, which would allow for a cleaner injection of the mock into the service under test.

Reply:
Moved away from using a mockable generator altogether for now, but yeah, that would have made the most sense.